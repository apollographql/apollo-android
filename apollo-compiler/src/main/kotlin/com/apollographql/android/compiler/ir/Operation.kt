package com.apollographql.android.compiler.ir

import com.apollographql.android.compiler.SchemaTypeSpecBuilder
import com.apollographql.apollo.api.Operation
import com.squareup.javapoet.TypeSpec
import javax.lang.model.element.Modifier

data class Operation(
    val operationName: String,
    val operationType: String,
    val variables: List<Variable>,
    val source: String,
    val fields: List<Field>,
    val filePath: String,
    val fragmentsReferenced: List<String>
) : CodeGenerator {
  override fun toTypeSpec(context: CodeGenerationContext): TypeSpec =
      SchemaTypeSpecBuilder(DATA_TYPE_NAME, fields, emptyList(), emptyList(), context)
          .build(Modifier.PUBLIC, Modifier.STATIC)
          .toBuilder()
          .addSuperinterface(Operation.Data::class.java)
          .build()

  companion object {
    val DATA_TYPE_NAME = "Data"
  }
}
