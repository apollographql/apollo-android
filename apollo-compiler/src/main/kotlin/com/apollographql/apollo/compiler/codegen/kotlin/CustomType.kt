package com.apollographql.apollo.compiler.codegen.kotlin

import com.apollographql.apollo.api.ScalarType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy

fun KotlinCodeGen.customScalarTypeSpec(customTypes: Map<String, String>): TypeSpec {
  return TypeSpec.enumBuilder("CustomType")
      .addAnnotation(generatedByApolloAnnotation)
      .addSuperinterface(ScalarType::class.java)
      .apply {
        customTypes.map { (schemaType, customType) ->
          addEnumConstant(
              name = schemaType.normalizeGraphQLType().toUpperCase(),
              typeSpec = enumConstantTypeSpec(schemaType = schemaType, customType = customType)
          )
        }
      }
      .build()
}

private fun enumConstantTypeSpec(schemaType: String, customType: String): TypeSpec {
  return TypeSpec.anonymousClassBuilder()
      .addFunction(FunSpec.builder("typeName")
          .addModifiers(KModifier.OVERRIDE)
          .returns(String::class)
          .addStatement("return %S", schemaType)
          .build()
      )
      .addFunction(FunSpec.builder("javaType")
          .returns(Class::class.asClassName().parameterizedBy(STAR))
          .addModifiers(KModifier.OVERRIDE)
          .addStatement("return %L::class.java", customType)
          .build()
      )
      .build()
}