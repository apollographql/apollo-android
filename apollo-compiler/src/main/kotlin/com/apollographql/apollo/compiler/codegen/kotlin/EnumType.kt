package com.apollographql.apollo.compiler.codegen.kotlin

import com.apollographql.apollo.compiler.applyIf
import com.apollographql.apollo.compiler.ast.AST
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeSpec

fun KotlinCodeGen.enumTypeSpec(enumType: AST.EnumType): TypeSpec {
  return with(enumType) {
    TypeSpec.enumBuilder(name)
        .applyIf(description.isNotBlank()) { addKdoc("%L\n", description) }
        .addAnnotation(generatedByApolloAnnotation)
        .primaryConstructor(primaryConstructorSpec)
        .addProperty(rawValuePropertySpec)
        .apply {
          values.forEach { value -> addEnumConstant(value.constName, value.enumConstTypeSpec) }
          addEnumConstant("UNKNOWN__", unknownEnumConstTypeSpec)
        }
        .addType(companionObjectSpec)
        .build()
  }
}

private val primaryConstructorSpec = FunSpec.constructorBuilder()
    .addParameter("rawValue", String::class)
    .build()

private val rawValuePropertySpec = PropertySpec.builder("rawValue", String::class)
    .initializer("rawValue")
    .build()

private val AST.EnumType.Value.enumConstTypeSpec: TypeSpec
  get() {
    return TypeSpec.anonymousClassBuilder()
        .applyIf(description.isNotBlank()) { addKdoc("%L\n", description) }
        .applyIf(isDeprecated) { addAnnotation(KotlinCodeGen.deprecatedAnnotation(deprecationReason)) }
        .addSuperclassConstructorParameter("%S", value)
        .build()
  }

private val unknownEnumConstTypeSpec: TypeSpec = TypeSpec.anonymousClassBuilder()
    .addKdoc("%L", "Auto generated constant for unknown enum values\n")
    .addSuperclassConstructorParameter("%S", "UNKNOWN__")
    .build()

private val AST.EnumType.companionObjectSpec: TypeSpec
  get() {
    return TypeSpec.companionObjectBuilder()
        .addFunction(safeValueOfFunSpec)
        .build()
  }

private val AST.EnumType.safeValueOfFunSpec: FunSpec
  get() {
    return FunSpec.builder("safeValueOf")
        .addAnnotation(JvmStatic::class)
        .addParameter("rawValue", String::class)
        .returns(ClassName("", name))
        .addCode("return values().find { it.rawValue == rawValue } ?: UNKNOWN__")
        .build()
  }