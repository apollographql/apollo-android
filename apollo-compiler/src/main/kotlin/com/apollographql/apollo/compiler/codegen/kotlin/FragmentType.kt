package com.apollographql.apollo.compiler.codegen.kotlin

import com.apollographql.apollo.api.GraphqlFragment
import com.apollographql.apollo.compiler.ast.AST
import com.apollographql.apollo.compiler.codegen.kotlin.KotlinCodeGen.toMapperFun
import com.apollographql.apollo.compiler.codegen.kotlin.KotlinCodeGen.responseFieldsPropertySpec
import com.apollographql.apollo.compiler.codegen.kotlin.KotlinCodeGen.asTypeName
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy

fun KotlinCodeGen.fragmentTypeSpec(fragmentType: AST.FragmentType): TypeSpec {
  return with(fragmentType) {
    TypeSpec.classBuilder(name)
      .addModifiers(KModifier.DATA)
      .addSuperinterface(GraphqlFragment::class.java)
      .addAnnotation(generatedByApolloAnnotation)
      .addAnnotation(suppressWarningsAnnotation)
      .primaryConstructor(primaryConstructorSpec)
      .addProperties(fields.map { field -> field.asPropertySpec(initializer = CodeBlock.of(field.name)) })
      .addType(companionObjectTypeSpec)
      .addTypes(nestedObjects.map { (_, type) -> objectTypeSpec(type) })
      .addFunction(marshallerFunSpec(fields).toBuilder().addModifiers(KModifier.OVERRIDE).build())
      .build()
  }
}

private val AST.FragmentType.primaryConstructorSpec: FunSpec
  get() {
    return FunSpec.constructorBuilder()
      .addParameters(fields.map { field ->
        val typeName = field.type.asTypeName()
        ParameterSpec.builder(
          name = field.name,
          type = if (field.isOptional) typeName.asNullable() else typeName
        ).build()
      })
      .build()
  }

private val AST.FragmentType.companionObjectTypeSpec: TypeSpec
  get() {
    return TypeSpec.companionObjectBuilder()
      .addProperty(responseFieldsPropertySpec(fields))
      .addProperty(PropertySpec.builder("FRAGMENT_DEFINITION", String::class)
        .initializer("%S", definition)
        .build()
      )
      .addProperty(PropertySpec.builder("POSSIBLE_TYPES",
        Array<String>::class.asClassName().parameterizedBy(String::class.asClassName()))
        .initializer(possibleTypes
          .map { CodeBlock.of("%S", it) }
          .joinToCode(prefix = "arrayOf(", separator = ", ", suffix = ")")
        )
        .build()
      )
      .addFunction(fields.toMapperFun(ClassName.bestGuess(name)))
      .build()
  }
