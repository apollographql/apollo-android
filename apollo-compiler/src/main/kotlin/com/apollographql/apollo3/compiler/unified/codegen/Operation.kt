package com.apollographql.apollo3.compiler.unified.codegen

import com.apollographql.apollo3.api.Mutation
import com.apollographql.apollo3.api.Query
import com.apollographql.apollo3.api.QueryDocumentMinifier
import com.apollographql.apollo3.api.Subscription
import com.apollographql.apollo3.compiler.backend.codegen.kotlinNameForOperation
import com.apollographql.apollo3.compiler.backend.codegen.makeDataClass
import com.apollographql.apollo3.compiler.unified.IrOperation
import com.apollographql.apollo3.compiler.unified.IrOperationType
import com.apollographql.apollo3.compiler.unified.codegen.helpers.maybeAddDescription
import com.apollographql.apollo3.compiler.unified.codegen.helpers.toNamedType
import com.apollographql.apollo3.compiler.unified.codegen.helpers.toParameterSpec
import com.apollographql.apollo3.compiler.unified.codegen.helpers.typeName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.asClassName

fun IrOperation.typeSpec(): TypeSpec {
  return TypeSpec.classBuilder(kotlinNameForOperation(name))
      .addSuperinterface(superInterfaceType())
      .maybeAddDescription(description)
      .makeDataClass(variables.map { it.toNamedType().toParameterSpec() })
      .addFunction(operationIdFunSpec())
      .addFunction(queryDocumentFunSpec())
      .addFunction(nameFunSpec())
      .addFunction(serializeVariablesFunSpec(packageName, name))
      .addFunction(adapterFunSpec(packageName = packageName, name = name, dataField = dataField))
      .addFunction(responseFieldsFunSpec())
      .addType(this.dataType.typeSpec(generateFragmentsAsInterfaces))
      .addType(companionTypeSpec())
      .build()
}

private fun IrOperation.superInterfaceType(): TypeName {
  return when (operationType) {
    IrOperationType.Query -> Query::class.asClassName()
    IrOperationType.Mutation -> Mutation::class.asClassName()
    IrOperationType.Subscription -> Subscription::class.asClassName()
  }.parameterizedBy(dataField.baseFieldSet!!.fullPath.typeName())
}

private fun operationIdFunSpec() = FunSpec.builder("operationId")
    .addModifiers(KModifier.OVERRIDE)
    .returns(String::class)
    .addStatement("return OPERATION_ID")
    .build()

private fun queryDocumentFunSpec() = FunSpec.builder("queryDocument")
    .addModifiers(KModifier.OVERRIDE)
    .returns(String::class)
    .addStatement("return QUERY_DOCUMENT")
    .build()

private fun nameFunSpec() = FunSpec.builder("name")
    .addModifiers(KModifier.OVERRIDE)
    .returns(String::class)
    .addStatement("return OPERATION_NAME")
    .build()

private fun IrOperation.companionTypeSpec(): TypeSpec {
  return TypeSpec.companionObjectBuilder()
      .addProperty(PropertySpec.builder("OPERATION_ID", String::class)
          .addModifiers(KModifier.CONST)
          .initializer("%S", )
          .build()
      )
      .addProperty(PropertySpec.builder("QUERY_DOCUMENT", String::class)
          .initializer(
              CodeBlock.builder()
                  .add("%T.minify(\n", QueryDocumentMinifier::class.java)
                  .indent()
                  .add("%S\n", sourceWithFragments)
                  .unindent()
                  .add(")")
                  .build()
          )
          .build()
      )
      .addProperty(PropertySpec
          .builder("OPERATION_NAME", String::class)
          .initializer("%S", name)
          .build()
      )
      .build()
}