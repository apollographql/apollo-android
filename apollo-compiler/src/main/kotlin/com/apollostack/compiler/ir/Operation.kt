package com.apollostack.compiler.ir

class Operation(val operationName: String, val operationType: String, val variables: List<String>,
    val source: String, val fields: List<Field>) {

}