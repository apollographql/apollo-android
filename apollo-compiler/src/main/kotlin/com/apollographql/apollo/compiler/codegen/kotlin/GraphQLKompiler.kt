package com.apollographql.apollo.compiler.codegen.kotlin

import com.apollographql.apollo.compiler.ClassNames
import com.apollographql.apollo.compiler.ast.toAST
import com.apollographql.apollo.compiler.formatPackageName
import com.apollographql.apollo.compiler.ir.CodeGenerationIR
import com.apollographql.apollo.compiler.ir.ScalarType
import com.apollographql.apollo.compiler.ir.TypeDeclaration
import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.asClassName
import com.squareup.moshi.Moshi
import java.io.File

class GraphQLKompiler(
    private val irFile: File,
    private val customTypeMap: Map<String, String>,
    private val outputPackageName: String?,
    private val useSemanticNaming: Boolean
) {
  private val moshi = Moshi.Builder().build()
  private val irAdapter = moshi.adapter(CodeGenerationIR::class.java)

  fun write(outputDir: File) {
    val ir = irAdapter.fromJson(irFile.readText())!!
    val irPackageName = outputPackageName ?: irFile.absolutePath.formatPackageName()
    val fragmentsPackage = if (irPackageName.isNotEmpty()) "$irPackageName.fragment" else "fragment"
    val typesPackageName = if (irPackageName.isNotEmpty()) "$irPackageName.type" else "type"
    val customTypeMap = customTypeMap.supportedTypeMap(ir.typesUsed)
    val ast = ir.toAST(
        customTypeMap = customTypeMap,
        typesPackageName = typesPackageName,
        fragmentsPackage = fragmentsPackage,
        useSemanticNaming = useSemanticNaming
    )

    if (irPackageName.isNotEmpty()) {
      File(outputDir, irPackageName.replace('.', File.separatorChar)).deleteRecursively()
    }

    if (ast.customTypes.isNotEmpty()) {
      KotlinCodeGen.customScalarTypeSpec(ast.customTypes).writeTo(outputDir, typesPackageName)
    }

    if (ast.enums.isNotEmpty()) {
      ast.enums.map { KotlinCodeGen.enumTypeSpec(it) }.forEach { it.writeTo(outputDir, typesPackageName) }
    }

    if (ast.inputTypes.isNotEmpty()) {
      ast.inputTypes.map { KotlinCodeGen.inputTypeSpec(it) }.forEach { it.writeTo(outputDir, typesPackageName) }
    }

    if (ast.fragments.isNotEmpty()) {
      ast.fragments.map { KotlinCodeGen.fragmentTypeSpec(it) }.forEach { it.writeTo(outputDir, fragmentsPackage) }
    }

    if (ast.operations.isNotEmpty()) {
      ast.operations.forEach { operation ->
        val targetPackage = outputPackageName ?: operation.filePath.formatPackageName()
        KotlinCodeGen.operationTypeSpec(operationType = operation, targetPackage = targetPackage).writeTo(
            outputDir = outputDir,
            packageName = targetPackage
        )
      }
    }
  }

  private fun TypeSpec.writeTo(outputDir: File, packageName: String) {
    FileSpec.builder(packageName, name!!)
        .addType(this)
        .build()
        .writeTo(outputDir)
  }

  private fun Map<String, String>.supportedTypeMap(typeDeclarations: List<TypeDeclaration>): Map<String, String> {
    val idScalarTypeMap = ScalarType.ID.name to (this[ScalarType.ID.name] ?: String::class.asClassName().toString())
    return typeDeclarations.filter { it.kind == TypeDeclaration.KIND_SCALAR_TYPE }
        .associate { it.name to (this[it.name] ?: ClassNames.OBJECT.toString()) }
        .plus(idScalarTypeMap)
  }
}