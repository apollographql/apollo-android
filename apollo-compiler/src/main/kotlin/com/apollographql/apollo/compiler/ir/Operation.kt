package com.apollographql.apollo.compiler.ir

import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.internal.SimpleResponseWriter
import com.apollographql.apollo.compiler.Annotations
import com.apollographql.apollo.compiler.ClassNames
import com.apollographql.apollo.compiler.SchemaTypeSpecBuilder
import com.apollographql.apollo.compiler.withBuilder
import com.apollographql.apollo.response.ScalarTypeAdapters
import com.squareup.javapoet.MethodSpec
import com.squareup.javapoet.ParameterSpec
import com.squareup.javapoet.TypeSpec
import java.io.IOException
import javax.lang.model.element.Modifier

data class Operation(
    val operationName: String,
    val operationType: String,
    val variables: List<Variable>,
    val source: String,
    val sourceWithFragments: String,
    val fields: List<Field>,
    val filePath: String,
    val fragmentsReferenced: List<String>
) : CodeGenerator {

  override fun toTypeSpec(context: CodeGenerationContext, abstract: Boolean): TypeSpec =
      SchemaTypeSpecBuilder(
          typeName = DATA_TYPE_NAME,
          fields = fields,
          fragmentRefs = emptyList(),
          inlineFragments = emptyList(),
          context = context,
          abstract = abstract
      )
          .build(Modifier.PUBLIC, Modifier.STATIC)
          .toBuilder()
          .addSuperinterface(Operation.Data::class.java)
          .addMethod(toJsonWithDefaultScalarTypesMethodSpec())
          .addMethod(toJsonMethodSpec())
          .build()
          .let {
            if (context.generateModelBuilder) {
              it.withBuilder()
            } else {
              it
            }
          }

  fun normalizedOperationName(useSemanticNaming: Boolean): String = when (operationType) {
    TYPE_MUTATION -> normalizedOperationName(useSemanticNaming, "Mutation")
    TYPE_QUERY -> normalizedOperationName(useSemanticNaming, "Query")
    TYPE_SUBSCRIPTION -> normalizedOperationName(useSemanticNaming, "Subscription")
    else -> throw IllegalArgumentException("Unknown operation type $operationType")
  }

  private fun normalizedOperationName(useSemanticNaming: Boolean, operationNameSuffix: String): String {
    return if (useSemanticNaming && !operationName.endsWith(operationNameSuffix)) {
      operationName.capitalize() + operationNameSuffix
    } else {
      operationName.capitalize()
    }
  }

  fun isMutation() = operationType == TYPE_MUTATION

  fun isQuery() = operationType == TYPE_QUERY

  fun isSubscription() = operationType == TYPE_SUBSCRIPTION

  private fun toJsonWithDefaultScalarTypesMethodSpec(): MethodSpec {
    return MethodSpec
        .methodBuilder("toJson")
        .addAnnotation(Annotations.OVERRIDE)
        .addModifiers(Modifier.PUBLIC)
        .returns(String::class.java)
        .addParameter(ParameterSpec
            .builder(ClassNames.STRING, "indent")
            .addAnnotation(Annotations.NONNULL)
            .build()
        )
        .addStatement("return toJson(indent, \$T.DEFAULT)", ScalarTypeAdapters::class.java)
        .build()
  }

  private fun toJsonMethodSpec(): MethodSpec {
    return MethodSpec
        .methodBuilder("toJson")
        .addAnnotation(Annotations.OVERRIDE)
        .addModifiers(Modifier.PUBLIC)
        .returns(String::class.java)
        .addParameter(ParameterSpec
            .builder(ClassNames.STRING, "indent")
            .addAnnotation(Annotations.NONNULL)
            .build()
        )
        .addParameter(ParameterSpec
            .builder(ScalarTypeAdapters::class.java, "scalarTypeAdapters")
            .addAnnotation(Annotations.NONNULL)
            .build()
        )
        .addCode("""
            try {
              final ${'$'}T responseWriter = new ${'$'}T(scalarTypeAdapters);
              marshaller().marshal(responseWriter);
              return responseWriter.toJson(indent);
            } catch (${'$'}T e) {
              throw new ${'$'}T(e);
            }
           
        """.trimIndent(), SimpleResponseWriter::class.java, SimpleResponseWriter::class.java, IOException::class.java,
            RuntimeException::class.java)
        .build()
  }

  companion object {
    const val DATA_TYPE_NAME = "Data"
    const val TYPE_MUTATION = "mutation"
    const val TYPE_QUERY = "query"
    const val TYPE_SUBSCRIPTION = "subscription"
  }
}
