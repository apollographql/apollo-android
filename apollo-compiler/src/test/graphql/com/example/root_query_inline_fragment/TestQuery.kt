// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package com.example.root_query_inline_fragment

import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.OperationName
import com.apollographql.apollo.api.Query
import com.apollographql.apollo.api.Response
import com.apollographql.apollo.api.ScalarTypeAdapters
import com.apollographql.apollo.api.ScalarTypeAdapters.Companion.DEFAULT
import com.apollographql.apollo.api.internal.OperationRequestBodyComposer
import com.apollographql.apollo.api.internal.QueryDocumentMinifier
import com.apollographql.apollo.api.internal.ResponseFieldMapper
import com.apollographql.apollo.api.internal.ResponseFieldMarshaller
import com.apollographql.apollo.api.internal.SimpleOperationResponseParser
import com.apollographql.apollo.api.internal.Throws
import com.example.root_query_inline_fragment.type.Episode
import kotlin.Boolean
import kotlin.Double
import kotlin.String
import kotlin.Suppress
import kotlin.collections.List
import okio.Buffer
import okio.BufferedSource
import okio.ByteString
import okio.IOException

@Suppress("NAME_SHADOWING", "UNUSED_ANONYMOUS_PARAMETER", "LocalVariableName",
    "RemoveExplicitTypeArguments", "NestedLambdaShadowedImplicitParameter", "PropertyName",
    "RemoveRedundantQualifierName")
class TestQuery : Query<TestQuery.Data, Operation.Variables> {
  override fun operationId(): String = OPERATION_ID

  override fun queryDocument(): String = QUERY_DOCUMENT

  override fun variables(): Operation.Variables = Operation.EMPTY_VARIABLES

  override fun name(): OperationName = OPERATION_NAME

  override fun responseFieldMapper(): ResponseFieldMapper<Data> {
    return ResponseFieldMapper { reader ->
      TestQuery_ResponseAdapter.fromResponse(reader)
    }
  }

  @Throws(IOException::class)
  override fun parse(source: BufferedSource, scalarTypeAdapters: ScalarTypeAdapters):
      Response<Data> {
    return SimpleOperationResponseParser.parse(source, this, scalarTypeAdapters)
  }

  @Throws(IOException::class)
  override fun parse(byteString: ByteString, scalarTypeAdapters: ScalarTypeAdapters):
      Response<Data> {
    return parse(Buffer().write(byteString), scalarTypeAdapters)
  }

  @Throws(IOException::class)
  override fun parse(source: BufferedSource): Response<Data> {
    return parse(source, DEFAULT)
  }

  @Throws(IOException::class)
  override fun parse(byteString: ByteString): Response<Data> {
    return parse(byteString, DEFAULT)
  }

  override fun composeRequestBody(scalarTypeAdapters: ScalarTypeAdapters): ByteString {
    return OperationRequestBodyComposer.compose(
      operation = this,
      autoPersistQueries = false,
      withQueryDocument = true,
      scalarTypeAdapters = scalarTypeAdapters
    )
  }

  override fun composeRequestBody(): ByteString = OperationRequestBodyComposer.compose(
    operation = this,
    autoPersistQueries = false,
    withQueryDocument = true,
    scalarTypeAdapters = DEFAULT
  )

  override fun composeRequestBody(
    autoPersistQueries: Boolean,
    withQueryDocument: Boolean,
    scalarTypeAdapters: ScalarTypeAdapters
  ): ByteString = OperationRequestBodyComposer.compose(
    operation = this,
    autoPersistQueries = autoPersistQueries,
    withQueryDocument = withQueryDocument,
    scalarTypeAdapters = scalarTypeAdapters
  )

  /**
   * The query type, represents all of the entry points into our object graph
   */
  interface Data : Operation.Data {
    fun asQuery(): Query? = this as? Query

    override fun marshaller(): ResponseFieldMarshaller

    /**
     * The query type, represents all of the entry points into our object graph
     */
    interface Query : Data {
      val __typename: String

      val hero: Hero?

      val droid: Droid?

      override fun marshaller(): ResponseFieldMarshaller

      /**
       * A character from the Star Wars universe
       */
      interface Hero {
        val __typename: String

        /**
         * The name of the character
         */
        val name: String

        /**
         * The movies this character appears in
         */
        val appearsIn: List<Episode?>

        fun marshaller(): ResponseFieldMarshaller

        /**
         * A humanoid creature from the Star Wars universe
         */
        interface Human : Hero {
          override val __typename: String

          /**
           * The name of the character
           */
          override val name: String

          /**
           * The movies this character appears in
           */
          override val appearsIn: List<Episode?>

          /**
           * Height in the preferred unit, default is meters
           */
          val height: Double?

          override fun marshaller(): ResponseFieldMarshaller
        }
      }

      /**
       * An autonomous mechanical character in the Star Wars universe
       */
      interface Droid {
        val __typename: String

        fun marshaller(): ResponseFieldMarshaller

        /**
         * An autonomous mechanical character in the Star Wars universe
         */
        interface Droid : Query.Droid {
          override val __typename: String

          /**
           * What others call this droid
           */
          val name: String

          /**
           * This droid's primary function
           */
          val primaryFunction: String?

          override fun marshaller(): ResponseFieldMarshaller
        }
      }
    }

    /**
     * The query type, represents all of the entry points into our object graph
     */
    data class QueryDatum(
      override val __typename: String = "Query",
      override val hero: Hero?,
      override val droid: Droid?
    ) : Data, Query {
      override fun marshaller(): ResponseFieldMarshaller {
        return ResponseFieldMarshaller { writer ->
          TestQuery_ResponseAdapter.Data.QueryDatum.toResponse(writer, this)
        }
      }

      /**
       * A character from the Star Wars universe
       */
      interface Hero : Query.Hero {
        override val __typename: String

        /**
         * The name of the character
         */
        override val name: String

        /**
         * The movies this character appears in
         */
        override val appearsIn: List<Episode?>

        fun asHuman(): Human? = this as? Human

        override fun marshaller(): ResponseFieldMarshaller

        /**
         * A humanoid creature from the Star Wars universe
         */
        interface Human : Query.Hero, Query.Hero.Human, Hero {
          override val __typename: String

          /**
           * The name of the character
           */
          override val name: String

          /**
           * The movies this character appears in
           */
          override val appearsIn: List<Episode?>

          /**
           * Height in the preferred unit, default is meters
           */
          override val height: Double?

          override fun marshaller(): ResponseFieldMarshaller
        }

        /**
         * A humanoid creature from the Star Wars universe
         */
        data class HumanHero(
          override val __typename: String = "Human",
          /**
           * The name of the character
           */
          override val name: String,
          /**
           * The movies this character appears in
           */
          override val appearsIn: List<Episode?>,
          /**
           * Height in the preferred unit, default is meters
           */
          override val height: Double?
        ) : Query.Hero, Query.Hero.Human, Hero {
          override fun marshaller(): ResponseFieldMarshaller {
            return ResponseFieldMarshaller { writer ->
              TestQuery_ResponseAdapter.Data.QueryDatum.Hero.HumanHero.toResponse(writer, this)
            }
          }
        }

        /**
         * A character from the Star Wars universe
         */
        data class OtherHero(
          override val __typename: String = "Character",
          /**
           * The name of the character
           */
          override val name: String,
          /**
           * The movies this character appears in
           */
          override val appearsIn: List<Episode?>
        ) : Query.Hero, Hero {
          override fun marshaller(): ResponseFieldMarshaller {
            return ResponseFieldMarshaller { writer ->
              TestQuery_ResponseAdapter.Data.QueryDatum.Hero.OtherHero.toResponse(writer, this)
            }
          }
        }
      }

      /**
       * An autonomous mechanical character in the Star Wars universe
       */
      interface Droid : Query.Droid {
        override val __typename: String

        fun asDroid(): Droid? = this as? Droid

        override fun marshaller(): ResponseFieldMarshaller

        /**
         * An autonomous mechanical character in the Star Wars universe
         */
        interface Droid : Query.Droid, Query.Droid.Droid, QueryDatum.Droid {
          override val __typename: String

          /**
           * What others call this droid
           */
          override val name: String

          /**
           * This droid's primary function
           */
          override val primaryFunction: String?

          override fun marshaller(): ResponseFieldMarshaller
        }

        /**
         * An autonomous mechanical character in the Star Wars universe
         */
        data class DroidDroid(
          override val __typename: String = "Droid",
          /**
           * What others call this droid
           */
          override val name: String,
          /**
           * This droid's primary function
           */
          override val primaryFunction: String?
        ) : Query.Droid, Query.Droid.Droid, QueryDatum.Droid {
          override fun marshaller(): ResponseFieldMarshaller {
            return ResponseFieldMarshaller { writer ->
              TestQuery_ResponseAdapter.Data.QueryDatum.Droid.DroidDroid.toResponse(writer, this)
            }
          }
        }

        /**
         * An autonomous mechanical character in the Star Wars universe
         */
        data class OtherDroid(
          override val __typename: String = "Droid"
        ) : Query.Droid, QueryDatum.Droid {
          override fun marshaller(): ResponseFieldMarshaller {
            return ResponseFieldMarshaller { writer ->
              TestQuery_ResponseAdapter.Data.QueryDatum.Droid.OtherDroid.toResponse(writer, this)
            }
          }
        }
      }
    }

    /**
     * The query type, represents all of the entry points into our object graph
     */
    class OtherDatum : Data {
      override fun marshaller(): ResponseFieldMarshaller {
        return ResponseFieldMarshaller { writer ->
          TestQuery_ResponseAdapter.Data.OtherDatum.toResponse(writer, this)
        }
      }
    }
  }

  companion object {
    const val OPERATION_ID: String =
        "22e04d7f32ad56e49eb2092e14a88853689bab52269dcc5d4d861d2d0a449657"

    val QUERY_DOCUMENT: String = QueryDocumentMinifier.minify(
          """
          |query TestQuery {
          |  ... on Query {
          |    __typename
          |    hero {
          |      __typename
          |      name
          |      appearsIn
          |      ... on Human {
          |        height
          |      }
          |    }
          |    droid(id: 1) {
          |      __typename
          |      ... on Droid {
          |        name
          |        primaryFunction
          |      }
          |    }
          |  }
          |}
          """.trimMargin()
        )

    val OPERATION_NAME: OperationName = object : OperationName {
      override fun name(): String {
        return "TestQuery"
      }
    }
  }
}
