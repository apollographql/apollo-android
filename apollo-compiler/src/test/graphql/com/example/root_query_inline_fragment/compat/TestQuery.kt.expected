//
// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL version '$VERSION'.
//
package com.example.root_query_inline_fragment

import com.apollographql.apollo3.api.Adapter
import com.apollographql.apollo3.api.CustomScalarAdapters
import com.apollographql.apollo3.api.FieldSet
import com.apollographql.apollo3.api.Query
import com.apollographql.apollo3.api.json.JsonWriter
import com.apollographql.apollo3.api.obj
import com.example.root_query_inline_fragment.adapter.TestQuery_ResponseAdapter
import com.example.root_query_inline_fragment.responsefields.TestQuery_ResponseFields
import com.example.root_query_inline_fragment.type.Episode
import kotlin.Double
import kotlin.String
import kotlin.collections.List

class TestQuery : Query<TestQuery.Data> {
  override fun id(): String = OPERATION_ID

  override fun document(): String = OPERATION_DOCUMENT

  override fun name(): String = OPERATION_NAME

  override fun serializeVariables(writer: JsonWriter, customScalarAdapters: CustomScalarAdapters) {
    // This operation doesn't have any variable
  }

  override fun adapter(): Adapter<Data> {
    return TestQuery_ResponseAdapter.Data.obj()
  }

  override fun fieldSets(): List<FieldSet> = TestQuery_ResponseFields.fields.first().fieldSets

  data class Data(
    val __typename: String,
    /**
     * Synthetic field for inline fragment on Query
     */
    val asQuery: AsQuery?
  ) : Query.Data

  data class AsHuman(
    /**
     * Height in the preferred unit, default is meters
     */
    val height: Double?,
    val __typename: String,
    /**
     * What this human calls themselves
     */
    val name: String,
    /**
     * The movies this human appears in
     */
    val appearsIn: List<Episode?>
  ) {
    fun appearsInFilterNotNull(): List<Episode> = appearsIn.filterNotNull()
  }

  data class Hero(
    val __typename: String,
    /**
     * The name of the character
     */
    val name: String,
    /**
     * The movies this character appears in
     */
    val appearsIn: List<Episode?>,
    /**
     * Synthetic field for inline fragment on Human
     */
    val asHuman: AsHuman?
  ) {
    fun appearsInFilterNotNull(): List<Episode> = appearsIn.filterNotNull()
  }

  data class AsDroid(
    /**
     * What others call this droid
     */
    val name: String,
    /**
     * This droid's primary function
     */
    val primaryFunction: String?,
    val __typename: String
  )

  data class Droid(
    val __typename: String,
    /**
     * Synthetic field for inline fragment on Droid
     */
    val asDroid: AsDroid?
  )

  data class AsQuery(
    val __typename: String,
    val hero: Hero?,
    val droid: Droid?
  )

  companion object {
    const val OPERATION_ID: String =
        "25584d760eab0f41189b9f2bbdbba3c0ec491aced65ef23924ecdc8f41ffe78c"

    /**
     * The minimized GraphQL document being sent to the server to save a few bytes.
     * The un-minimized version is:
     *
     * query TestQuery {
     *   __typename
     *   ... on Query {
     *     __typename
     *     hero {
     *       __typename
     *       name
     *       appearsIn
     *       ... on Human {
     *         height
     *       }
     *     }
     *     droid(id: 1) {
     *       __typename
     *       ... on Droid {
     *         name
     *         primaryFunction
     *       }
     *     }
     *   }
     * }
     */
    const val OPERATION_DOCUMENT: String =
        "query TestQuery { __typename ... on Query { __typename hero { __typename name appearsIn ... on Human { height } } droid(id: 1) { __typename ... on Droid { name primaryFunction } } } }"

    const val OPERATION_NAME: String = "TestQuery"
  }
}
